File: leetcode/20190219/max-area-of-island.andi.java
class Solution {
  public int maxAreaOfIsland(int[][] grid) {
      int maxArea = 0;
      for (int i = 0; i < grid.length; i++) {
          for (int j = 0; j < grid[i].length; j++) {
              final int markedArea = markArea(grid, i, j);
              if (markedArea > maxArea) {
                  maxArea = markedArea;
              }
          }
      }
      return maxArea;
  }
  
  private int markArea(int[][] grid, int i, int j) {
      if (i >= grid.length || i < 0 || j >= grid[i].length || j < 0) {
          return 0;
      }
      if (grid[i][j] == 0) {
          return 0;
      }
      grid[i][j] = 0;
      return 1 + markArea(grid, i + 1, j) + markArea(grid, i - 1, j)
          + markArea(grid, i, j + 1) + markArea(grid, i, j - 1);
  }
}

File: leetcode/20190219/max-area-of-island.andi.py
class Solution:
    def maxAreaOfIsland(self, grid: 'List[List[int]]') -> 'int':
        max_area = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                tiles = self.mark_area(grid, i, j)
                area = len(tiles)
                if area > max_area:
                    max_area = area
        return max_area
    def mark_area(self, grid: 'List[List[int]]', i: int, j: int) -> 'int':
        if i >= len(grid) or i < 0 or j >= len(grid[i]) or j < 0:
            return []
        if grid[i][j] == 0:
            return []
        
        tiles = [(i, j)]
        grid[i][j] = 0
        tiles += self.mark_area(grid, i + 1, j)
        tiles += self.mark_area(grid, i - 1, j)
        tiles += self.mark_area(grid, i, j + 1)
        tiles += self.mark_area(grid, i, j - 1)
        
        return tiles